### PR TITLE
ref(auth): Have AuthHelper extend Pipeline

### PR DESCRIPTION
**Pull request by RyanSkonnord**
https://getsentry.atlassian.net/browse/ER-707

### CODE REVIEW

File: src/sentry/auth/helper.py [hunk 1]
```diff
@@ -56,62 +57,14 @@
...
-    # Expire the pipeline after 10 minutes of inactivity.
-    EXPIRATION_TTL = 10 * 60
-
-    def __init__(self, request):
-        self.__dict__["request"] = request
-
-    @property
-    def _client(self):
-        return clusters.get("default").get_local_client_for_key(self.auth_key)
+class RedisBackedState(PipelineSessionStore):
```
Comment (EvanPurkhiser) [inline src/sentry/auth/helper.py:60]: Hmm, I thought there was a way to use this without having to extend it. Maybe not?

Comment (EvanPurkhiser) [inline src/sentry/auth/helper.py:60]: I would probably name it something else though ideally

File: src/sentry/utils/session_store.py [hunk 2]
```diff
@@ -13,6 +13,13 @@ class RedisSessionStore:
     the request session. Useful for storing data too large to be stored into
     the session cookie.
 
+    The attributes to be backed by Redis must be declared in a subclass using
+    the `redis_proerty` function. For example:
```
Comment (RyanSkonnord) [inline src/sentry/utils/session_store.py:17]: Changing from dynamic attributes was motivated primarily by weird errors I kept hitting while testing, where the `redis_key` or `auth_key` or something would go to the `__getattr__` method (thereby causing infinite recursion) instead of to the property. Even if those bugs are avoidable, I would defend this change.

**Pro:**

- Avoids surprising differences between Redis-backed attributes from regular attributes.
- Exposes intent up front of what is to be stored in Redis.
- Allows IDEs to see and jump to (e.g., Cmd+Click) a declared property definition.
- Prevents risk of bugs from `__getattr__` and `__setattr__` being hit unintentionally.

**Con:**

- Necessitates subclasses to add properties for individual use cases.
- Existing references to attributes not covered this way will cause `AttributeError` (but tests should catch all of those).

We could avoid the subclasses by pulling the full universe of properties up into the base class, but this class should be a general tool and I don't think concerns about the attributes to be stored (e.g., "flow") belong up here.

File: src/sentry/auth/helper.py [hunk 3]
```diff
@@ -551,104 +504,64 @@ class AuthHelper:
...
-        organization = Organization.objects.get(id=state.org_id)
+        flow = req_state.state.flow
 
         return cls(
-            request, organization, flow, auth_provider=auth_provider, provider_key=provider_key
+            request,
+            req_state.organization,
+            flow,
+            auth_provider=req_state.provider_model,
+            provider_key=req_state.provider_key,
```
Comment (RyanSkonnord) [inline src/sentry/auth/helper.py:528]: `get_for_request` is the one of a few `Pipeline` methods that `AuthHelper` overrides entirely, rather than dispatching to `super`. This is a concession to having a completely different `__init__` signature; neither `__init__` method is changed.

I extracted the new `unpack_state` method to avoid as much code duplication as possible.

File: src/sentry/auth/helper.py [hunk 4]
```diff
@@ -551,104 +504,64 @@ class AuthHelper:
...
-        return self.pipeline[step_index].dispatch(request=self.request, helper=self)
-
-    def next_step(self):
-        """
-        Render the next step.
-        """
-        self.state.step_index += 1
-        return self.current_step()
+    def dispatch_to(self, step: View):
+        return step.dispatch(request=self.request, helper=self)
```
Comment (RyanSkonnord) [inline src/sentry/auth/helper.py:564]: This is a weird little kludge. The overridden method calls the second kwarg `pipeline` instead of `helper`. AFIACT, that's a hard-coded parameter name in a lot of existing view classes. An overridden `dispatch_to` method seemed preferable to changing them all.

File: src/sentry/auth/helper.py (hunk 1 above)
Comment (RyanSkonnord) [reply src/sentry/auth/helper.py:60]: The complications are:

1. The original code constructed `RedisBackedState` in two places: `get_for_request_` and `__init__`. Even if we're overriding it in `get_for_request_`, `Pipeline.__init__` would still be assigning a `RedisSessionStore`. So they need to be unified somehow.
2. The original code was using different prefixes from `RedisSessionStore`. It *might* be safe to just change that, but I'd assume it risks breaking the flow is a user is midway through at the time the change is deployed. So the extra attributes let us preserve exact behavior.
3. If we go forward with `RedisSessionStore` no longer allowing dynamic attributes, this is the best place for `flow = redis_property("flow")` to live.

I would support renaming it; only left it alone to minimize code movement.

File: src/sentry/utils/session_store.py [hunk 5]
```diff
@@ -71,7 +80,11 @@ def clear(self):
             return
 
         self._client.delete(self.redis_key)
-        del self.request.session[self.session_key]
+
+        session = self.request.session
+        del session[self.session_key]
+        if hasattr(session, "modified"):
```
Comment (JoshFerge) [inline src/sentry/utils/session_store.py:86]: what is 
```
        if hasattr(session, "modified"):
            session.modified = True
```
doing?

Comment (RyanSkonnord) [reply src/sentry/utils/session_store.py:86]: Replicating https://github.com/getsentry/sentry/pull/26883/files#diff-e0494c56201878812ba2d58e03a3e6310b0da0b3ed3f5c8ee6c0c9fa243d54aaL89. To be honest, beyond that I don't know.

When I put this here, I assumed it was harmless, but the `hasattr` was because it turned out to break some other cases. Thanks for provoking a second look -- I'm going to add a new hook for this.

File: src/sentry/integrations/pipeline.py [hunk 6]
```diff
@@ -33,6 +35,10 @@ class IntegrationPipeline(Pipeline):
     pipeline_name = "integration_pipeline"
     provider_manager = default_manager
 
+    def get_analytics_entry(self) -> Optional[PipelineAnalyticsEntry]:
```
Comment (EvanPurkhiser) [inline src/sentry/integrations/pipeline.py:38]: Let's definitely try and pull this out into it's own PR for moving the analytics stuff

File: src/sentry/pipeline/__init__.py [hunk 7]
```diff
@@ -33,7 +37,13 @@ def set_config(self, config):
         the provider instance. This is useful for example when nesting
         pipelines and the provider needs to be configured differently.
         """
-        self.config = config
+        try:
```
Comment (EvanPurkhiser) [inline src/sentry/pipeline/__init__.py:40]: Yeah I think if we just ensure config is always set, we can get rid of this logic.

We should probably rename `set_config` to `update_config` too then

File: src/sentry/integrations/pipeline.py (hunk 6 above)
Comment (RyanSkonnord) [reply src/sentry/integrations/pipeline.py:38]: Done: https://github.com/getsentry/sentry/pull/27176

**Review by evanpurkhiser — APPROVED** [review]

**Comment by RyanSkonnord** [comment]
Thanks @EvanPurkhiser. Will deploy after a quick smoke test on Staging.
